url79d4f7c7ed_schanq_Live_Algorithm_tgz-pJ8-algorithms_and_systems_AI_Evolutionary_GeneticAlgorithmJ8/src/algorithms_and_systems_AI_Evolutionary/FuzzySystem.java

```diff
@@ -0,0 +1,938 @@
+			if(k.ordinal() == i){
+				theKey = k;
+				break;
+			}
+		}
+		// Then create arrays for the relative minor and majorthirds,
+		// tally the matches, sort the result and the rule that returns
+		// true determines the tonality
+		int firstOct = theKey.getOctave0();
+		theKey.setRelativeMinor();
+		int [] majorRoots_minorThirds 	= theKey.getOctaves();
+		int [] minorRoots_majorSixths 	= theKey.getRelativeMinor().getOctaves();
+		int [] majorThirds_minorTsFifth	= new int [allTwelveTonesMajorScales[theKey.ordinal()].length/7];
+		
+		matches = new int [3];
+		
+		if(firstOct == 0){
+			i = 0;
+			for(int j = firstOct+2; j < allTwelveTonesMajorScales[theKey.ordinal()].length; j+=7){
+				majorThirds_minorTsFifth[i] = allTwelveTonesMajorScales[theKey.ordinal()][j];
+				i++;
+			}
+		}
+		else{
+			i = 0;
+			for(int j = firstOct-5; j < allTwelveTonesMajorScales[theKey.ordinal()].length; j+=7){
+				if(j >= 0){
+					majorThirds_minorTsFifth[i] = allTwelveTonesMajorScales[theKey.ordinal()][j];
+					i++;
+				}
+			}
+		}
+		for(int row = 0; row < notes.length; row++){
+			for(i = 0; i < majorThirds_minorTsFifth.length; i++){
+				if(notes[row] == majorRoots_minorThirds[i])
+					matches[0]++;
+				if(notes[row] == minorRoots_majorSixths[i])
+					matches[1]++;
+				if(notes[row] == majorThirds_minorTsFifth[i])
+					matches[2]++;
+			}
+		}
+		int maj1_min3 = matches[0];
+		int maj6_min1 = matches[1];
+		int maj3_min5 = matches[2];
+		
+		this.sort(matches);
+		
+		if((maj1_min3 == matches[2] && maj6_min1 == matches[1])
+		 ||(maj1_min3 == matches[2] && maj3_min5 == matches[1])
+		 ||(maj3_min5 == matches[2] && maj1_min3 == matches[1]))
+			majmin = 1;
+		else if((maj6_min1 == matches[2] && maj3_min5 == matches[1])
+			  ||(maj6_min1 == matches[2] && maj1_min3 == matches[1])
+			  ||(maj3_min5 == matches[2] && maj6_min1 == matches[1]))
+			majmin = 0;
+		
+		return new Tonality(theKey, majmin);
+	}
+	
+	// ==========================================================
+	// =		METHODS TO DETERMINE TARGET LOUDNESS			=
+	// ==========================================================
+
+	// Equation for working out the slope of a line (used to	
+	// analyse whether a data set is on average ascending/		
+	// descending /flat as makes no difference					
+	private double slope(double x0, double y0, double x1, double y1){
+		
+		double m = (y1-y0)/(x1-x0);
+		return m;
+		
+	}
+	// Uses above slope equation to determine if there is a 	
+	// direction/slope to a data set.							
+	protected double averageDirection(int [] set){
+		
+		double summedSlope = 0;
+		
+		for(int i = 1; i < set.length-1; i++){
+			
+			summedSlope += this.slope(i, set[i], i+1, set[i+1]);
+			
+		}
+		return summedSlope/set.length;
+	}
+	// Method to determine the DynamicShape enum is most true	
+	// for the performers current playing state. The output	to	
+	// be fed into the fitness function of the Genetic Algorithm
+	private DynamicShape determineDynamicShape(int [] vels){
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This method contains 79 lines of code, not including blank lines or lines with only comments, Java punctuation characters, identifiers, or literals. By comparison, 99% of the methods in the CodeGuru reference dataset contain fewer lines of code. Large methods might be difficult to read and have logic that is hard to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 510-585 into a separate method.